File: software/ips_jazzy_ws/src/ips_nodes_cpp/src/differentiator_node.cpp
// =============================================================================
//  differentiator_node.cpp  (Tingkat 1.5 — Savitzky-Golay)
//
//    /state/position_compensated -> /state/translation_velocity
//    /state/angular_velocity     -> /state/angular_acceleration  (enable_angular)
//
//  Metode turunan dipilih via parameter `deriv_method`:
//    "backward" : beda-hingga mundur (lama; memperkuat noise)
//    "savgol"   : Savitzky-Golay causal (default; fit poly lalu turunkan,
//                 jauh lebih halus untuk sinyal noisy — port read_UDP.py)
//
//  SG memakai dt AKTUAL tiap sampel (bukan FIXED_DT 50Hz read_UDP.py), jadi
//  benar di rate IMU/posisi berapa pun. Window & orde poly via parameter.
//  Clip percepatan sudut opsional (read_UDP.py: MAX_ANGULAR_ACCEL).
// =============================================================================
#include <memory>
#include <string>

#include <rclcpp/rclcpp.hpp>

#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
#include "geometry_msgs/msg/vector3_stamped.hpp"

#include "ips_nodes_cpp/imu_filters.hpp"

namespace ips {

using std::placeholders::_1;

static rclcpp::QoS qos_state_reliable() {
    rclcpp::QoS q(20); q.reliable().keep_last(20); return q;
}

class DifferentiatorNode : public rclcpp::Node {
public:
    DifferentiatorNode() : rclcpp::Node("differentiator") {
        declare_parameter<double>("min_dt_s", 1e-4);
        declare_parameter<double>("max_dt_s", 1.0);
        declare_parameter<bool>  ("enable_angular", false);
        declare_parameter<std::string>("deriv_method", "savgol");  // "savgol"|"backward"
        declare_parameter<int>   ("sg_window", 15);
        declare_parameter<int>   ("sg_poly",   3);
        declare_parameter<double>("ang_accel_clip", 0.0);  // 0 = tanpa clip

        min_dt_         = get_parameter("min_dt_s").as_double();
        max_dt_         = get_parameter("max_dt_s").as_double();
        enable_angular_ = get_parameter("enable_angular").as_bool();
        method_         = get_parameter("deriv_method").as_string();
        const int win   = get_parameter("sg_window").as_int();
        const int poly  = get_parameter("sg_poly").as_int();
        const double clip = get_parameter("ang_accel_clip").as_double();

        use_sg_ = (method_ == "savgol");

        // SG instances: 3 utk velocity (pos), 3 utk angular accel
        for (int i = 0; i < 3; ++i) {
            sg_vel_[i]   = SavGolDerivative(static_cast<std::size_t>(win), poly, 0.0);
            sg_alpha_[i] = SavGolDerivative(static_cast<std::size_t>(win), poly, clip);
        }

        sub_pos_ = create_subscription<
            geometry_msgs::msg::PoseWithCovarianceStamped>(
            "/state/position_compensated", qos_state_reliable(),
            std::bind(&DifferentiatorNode::on_position, this, _1));
        pub_vel_ = create_publisher<geometry_msgs::msg::Vector3Stamped>(
            "/state/translation_velocity", qos_state_reliable());

        if (enable_angular_) {
            sub_omega_ = create_subscription<
                geometry_msgs::msg::Vector3Stamped>(
                "/state/angular_velocity", qos_state_reliable(),
                std::bind(&DifferentiatorNode::on_angular_velocity, this, _1));
            pub_alpha_ = create_publisher<geometry_msgs::msg::Vector3Stamped>(
                "/state/angular_acceleration", qos_state_reliable());
        }

        RCLCPP_INFO(get_logger(),
            "differentiator (C++) ready — method=%s angular=%s (sg win=%d poly=%d)",
            method_.c_str(), enable_angular_ ? "true" : "false", win, poly);
    }

private:
    void on_position(
        const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg) {
        const double t = stamp_to_sec(msg->header.stamp);
        const double x = msg->pose.pose.position.x;
        const double y = msg->pose.pose.position.y;
        const double z = msg->pose.pose.position.z;

        if (have_prev_pos_) {
            const double dt = t - pt_;
            if (dt >= min_dt_ && dt <= max_dt_) {
                double vx, vy, vz;
                if (use_sg_) {
                    vx = sg_vel_[0].update(x, dt);
                    vy = sg_vel_[1].update(y, dt);
                    vz = sg_vel_[2].update(z, dt);
                    if (!sg_vel_[0].ready()) { pt_=t; px_=x; py_=y; pz_=z; return; }
                } else {
                    vx = (x - px_) / dt; vy = (y - py_) / dt; vz = (z - pz_) / dt;
                }
                publish_vec(pub_vel_, msg->header.stamp, msg->header.frame_id, vx, vy, vz);
            }
        } else if (use_sg_) {
            sg_vel_[0].update(x, 0.05);  // seed window (dt nominal, hasil diabaikan)
            sg_vel_[1].update(y, 0.05);
            sg_vel_[2].update(z, 0.05);
        }
        pt_ = t; px_ = x; py_ = y; pz_ = z;
        have_prev_pos_ = true;
    }

    void on_angular_velocity(
        const geometry_msgs::msg::Vector3Stamped::SharedPtr msg) {
        const double t  = stamp_to_sec(msg->header.stamp);
        const double wx = msg->vector.x, wy = msg->vector.y, wz = msg->vector.z;

        if (have_prev_omega_) {
            const double dt = t - ot_;
            if (dt >= min_dt_ && dt <= max_dt_) {
                double ax, ay, az;
                if (use_sg_) {
                    ax = sg_alpha_[0].update(wx, dt);
                    ay = sg_alpha_[1].update(wy, dt);
                    az = sg_alpha_[2].update(wz, dt);
                    if (!sg_alpha_[0].ready()) { ot_=t; owx_=wx; owy_=wy; owz_=wz; return; }
                } else {
                    ax = (wx - owx_) / dt; ay = (wy - owy_) / dt; az = (wz - owz_) / dt;
                }
                publish_vec(pub_alpha_, msg->header.stamp, msg->header.frame_id, ax, ay, az);
            }
        } else if (use_sg_) {
            sg_alpha_[0].update(wx, 0.05);
            sg_alpha_[1].update(wy, 0.05);
            sg_alpha_[2].update(wz, 0.05);
        }
        ot_ = t; owx_ = wx; owy_ = wy; owz_ = wz;
        have_prev_omega_ = true;
    }

    static void publish_vec(
        const rclcpp::Publisher<geometry_msgs::msg::Vector3Stamped>::SharedPtr& pub,
        const builtin_interfaces::msg::Time& stamp,
        const std::string& frame_id, double x, double y, double z) {
        geometry_msgs::msg::Vector3Stamped out;
        out.header.stamp = stamp; out.header.frame_id = frame_id;
        out.vector.x = x; out.vector.y = y; out.vector.z = z;
        pub->publish(out);
    }

    static double stamp_to_sec(const builtin_interfaces::msg::Time& s) {
        return static_cast<double>(s.sec) + static_cast<double>(s.nanosec) * 1e-9;
    }

    double min_dt_ = 1e-4, max_dt_ = 1.0;
    bool   enable_angular_ = false, use_sg_ = true;
    std::string method_ = "savgol";

    SavGolDerivative sg_vel_[3], sg_alpha_[3];

    bool   have_prev_pos_ = false;
    double pt_=0, px_=0, py_=0, pz_=0;
    bool   have_prev_omega_ = false;
    double ot_=0, owx_=0, owy_=0, owz_=0;

    rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr sub_pos_;
    rclcpp::Subscription<geometry_msgs::msg::Vector3Stamped>::SharedPtr            sub_omega_;
    rclcpp::Publisher<geometry_msgs::msg::Vector3Stamped>::SharedPtr               pub_vel_;
    rclcpp::Publisher<geometry_msgs::msg::Vector3Stamped>::SharedPtr               pub_alpha_;
};

}  // namespace ips

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ips::DifferentiatorNode>());
    rclcpp::shutdown();
    return 0;
}

File: software/ips_jazzy_ws/src/ips_nodes/ips_nodes/differentiator_node.py
"""differentiator_node — backward-difference derivative.

At this stage of the project (no IMU yet), only translation velocity is
computed (from /state/position). Once the tag IMU is online and the
imu_processor publishes /state/angular_velocity, the angular_acceleration
branch automatically becomes active — set `enable_angular` to true.

Inputs:
    /state/position           (geometry_msgs/PoseWithCovarianceStamped, RELIABLE)
    /state/angular_velocity   (geometry_msgs/Vector3Stamped, RELIABLE) [optional]

Outputs:
    /state/translation_velocity (geometry_msgs/Vector3Stamped, RELIABLE)
    /state/angular_acceleration (geometry_msgs/Vector3Stamped, RELIABLE) [optional]

Note: backward difference amplifies noise. The position KF inside
position_solver already contains a velocity state, so a future
improvement is to publish that velocity directly and skip this node
for translation_velocity.
"""

from typing import Optional

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import (
    PoseWithCovarianceStamped,
    Vector3Stamped,
)

from ips_nodes.common import QOS_STATE_RELIABLE


def _stamp_to_seconds(stamp) -> float:
    return float(stamp.sec) + float(stamp.nanosec) * 1e-9


class DifferentiatorNode(Node):
    """Computes time-derivatives of state streams."""

    def __init__(self) -> None:
        super().__init__('differentiator')

        # ---- Parameters ------------------------------------------------
        self.declare_parameter('min_dt_s', 1e-4)
        self.declare_parameter('max_dt_s', 1.0)
        # IMU-side branch (off until tag IMU is integrated)
        self.declare_parameter('enable_angular', False)

        self._min_dt: float = float(self.get_parameter('min_dt_s').value)
        self._max_dt: float = float(self.get_parameter('max_dt_s').value)
        self._enable_angular: bool = bool(self.get_parameter('enable_angular').value)

        # ---- Position branch (always on) ------------------------------
        self._sub_pos = self.create_subscription(
            PoseWithCovarianceStamped,
            '/state/position_compensated',
            self._on_position,
            QOS_STATE_RELIABLE,
        )
        self._pub_vel = self.create_publisher(
            Vector3Stamped,
            '/state/translation_velocity',
            QOS_STATE_RELIABLE,
        )
        self._prev_pos: Optional[tuple[float, float, float, float]] = None

        # ---- Angular branch (optional) --------------------------------
        self._prev_omega: Optional[tuple[float, float, float, float]] = None
        if self._enable_angular:
            self._sub_omega = self.create_subscription(
                Vector3Stamped,
                '/state/angular_velocity',
                self._on_angular_velocity,
                QOS_STATE_RELIABLE,
            )
            self._pub_alpha = self.create_publisher(
                Vector3Stamped,
                '/state/angular_acceleration',
                QOS_STATE_RELIABLE,
            )

        self.get_logger().info(
            f'differentiator ready (angular={self._enable_angular})'
        )

    # ------------------------------------------------------------------
    def _on_position(self, msg: PoseWithCovarianceStamped) -> None:
        t = _stamp_to_seconds(msg.header.stamp)
        x = msg.pose.pose.position.x
        y = msg.pose.pose.position.y
        z = msg.pose.pose.position.z

        if self._prev_pos is not None:
            t0, x0, y0, z0 = self._prev_pos
            dt = t - t0
            if self._min_dt <= dt <= self._max_dt:
                vx = (x - x0) / dt
                vy = (y - y0) / dt
                vz = (z - z0) / dt
                self._publish_vector(
                    self._pub_vel,
                    msg.header.stamp,
                    msg.header.frame_id,
                    vx, vy, vz,
                )
        self._prev_pos = (t, x, y, z)

    def _on_angular_velocity(self, msg: Vector3Stamped) -> None:
        t = _stamp_to_seconds(msg.header.stamp)
        wx, wy, wz = msg.vector.x, msg.vector.y, msg.vector.z

        if self._prev_omega is not None:
            t0, wx0, wy0, wz0 = self._prev_omega
            dt = t - t0
            if self._min_dt <= dt <= self._max_dt:
                ax = (wx - wx0) / dt
                ay = (wy - wy0) / dt
                az = (wz - wz0) / dt
                self._publish_vector(
                    self._pub_alpha,
                    msg.header.stamp,
                    msg.header.frame_id,
                    ax, ay, az,
                )
        self._prev_omega = (t, wx, wy, wz)

    # ------------------------------------------------------------------
    @staticmethod
    def _publish_vector(pub, stamp, frame_id, x, y, z) -> None:
        out = Vector3Stamped()
        out.header.stamp = stamp
        out.header.frame_id = frame_id
        out.vector.x = float(x)
        out.vector.y = float(y)
        out.vector.z = float(z)
        pub.publish(out)


def main(args=None) -> None:
    rclpy.init(args=args)
    node = DifferentiatorNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == '__main__':
    main()
